## 🔢 Calculative Foundation


### 🛠️ import libraries

In [1]:
import pandas as pd
import numpy as np


### 📁 Read Excel file

In [2]:

df = pd.read_excel("Student_Performance.xlsx")
print("read excel file succesfully !")


read excel file succesfully !


In [3]:
df.head()

,Student_ID,Math,Physics,Chemistry,English,Computer,Category
0,S01,92,88,90,78,95,Above Average
1,S02,75,70,72,80,74,Below Average
2,S03,85,90,88,84,91,Above Average
3,S04,60,65,62,70,68,Below Average
4,S05,95,94,96,89,98,Above Average


In [4]:
df.tail()

,Student_ID,Math,Physics,Chemistry,English,Computer,Category
45,S46,77,79,78,80,76,Below Average
46,S47,87,85,88,84,90,Above Average
47,S48,56,58,57,64,61,Below Average
48,S49,83,84,82,81,86,Above Average
49,S50,65,67,66,71,68,Below Average


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Student_ID  50 non-null     object
 1   Math        50 non-null     int64 
 2   Physics     50 non-null     int64 
 3   Chemistry   50 non-null     int64 
 4   English     50 non-null     int64 
 5   Computer    50 non-null     int64 
 6   Category    50 non-null     object
dtypes: int64(5), object(2)
memory usage: 2.9+ KB


In [7]:
df.describe()

,Math,Physics,Chemistry,English,Computer
count,50.000000,50.000000,50.00000,50.000000,50.000000
mean,77.840000,78.840000,78.32000,78.240000,80.820000
std,12.851364,11.590566,12.64498,7.867681,12.379526
min,55.000000,58.000000,56.00000,63.000000,60.000000
25%,67.250000,69.250000,68.25000,72.000000,69.250000
50%,79.000000,80.500000,79.00000,79.000000,82.000000
75%,88.000000,88.750000,88.75000,84.000000,91.000000
max,99.000000,98.000000,99.00000,93.000000,100.000000


In [11]:
df.shape

(50, 7)

In [12]:
df.columns

Index(['Student_ID', 'Math', 'Physics', 'Chemistry', 'English', 'Computer',
       'Category'],
      dtype='object')

## Part A: Vector & Matrix Fundamentals


#### Q1. Represent each student’s subject scores as a vector.


In [13]:
# Select only subject columns
data = df[["Math", "Physics", "Chemistry", "English", "Computer"]].values

# Student vectors
v1 = data[0]   # S01
v2 = data[1]   # S02

print("Student 1 (S01) Vector:", v1)
print("Student 2 (S02) Vector:", v2)

Student 1 (S01) Vector: [92 88 90 78 95]
Student 2 (S02) Vector: [75 70 72 80 74]


**🎯 Insight:** Each student is now a **5-dimensional vector** in the subject space. S01 = high performer 🏆, S02 = average performer.

#### Q2. Compute:
#### (a) Norm-1 and Norm-2 of vectors.


In [14]:
l1 = np.linalg.norm(v1, ord=1)
l2 = np.linalg.norm(v1, ord=2)

print("\nL1 Norm:", l1)
print("L2 Norm:", l2)


L1 Norm: 443.0
L2 Norm: 198.5371501759809


**📏 Interpretation:**
- **L1 norm = 443** → total marks scored by S01 (sum of absolute values).
- **L2 norm ≈ 198.54** → geometric length of the score vector in 5-D space.

Both confirm S01 is a strong scorer overall. 🌟

#### (b) Dot product and angle between two students’ score vectors.



In [15]:
dot = np.dot(v1, v2)

angle = np.arccos(dot / (np.linalg.norm(v1) * np.linalg.norm(v2)))
angle_deg = np.degrees(angle)

print("\nDot Product:", dot)
print("Angle (Degrees):", angle_deg)


Dot Product: 32810
Angle (Degrees): 5.725561179228223


**📐 Insight:** Angle between S01 and S02 is only **≈ 5.73°** — extremely small ➡️ the two students’ score patterns point in almost the **same direction** in vector space 🎯. They differ in *magnitude* (S01 > S02) but not in *shape* of performance.

#### (c) Cross product (for 3D selected subjects).

In [16]:
A3 = df.loc[df["Student_ID"]=="S01",
            ["Math","Physics","Chemistry"]].values.flatten()

B3 = df.loc[df["Student_ID"]=="S02",
            ["Math","Physics","Chemistry"]].values.flatten()

cross_product = np.cross(A3,B3)

print("Student S01 :",A3)
print("Student S02 :",B3)

print("\nCross Product =")
print(cross_product)


Student S01 : [92 88 90]
Student S02 : [75 70 72]

Cross Product =
[  36  126 -160]


**✖️ Interpretation:** The cross product `[36, 126, -160]` is a vector **perpendicular** to both S01 and S02 in the Math–Physics–Chemistry space. A non-zero result confirms the two vectors are **not parallel** in 3D.

#### Q3. Find the projection of one vector onto another.

In [19]:
projection = (np.dot(v1, v2) / np.dot(v2, v2)) * v2

print("Projection of S01 onto S02")
print(np.round(projection,2))

Projection of S01 onto S02
[89.21 83.26 85.64 95.15 88.02]


**🎯 Insight:** The projection shows *how much of S01 lies along the direction of S02*. Since the projected vector `[89.21, 83.26, 85.64, 95.15, 88.02]` is very close in magnitude to S01’s own values, S01 and S02 share a **highly aligned performance direction**. 🔗

## Part B: Matrix Operations


#### Q4. form a matrix of students × subjects. Perform:



In [27]:
A = data
print("Matrix A (First 5 Students):")
print(A[:5])


Matrix A (First 5 Students):
[[92 88 90 78 95]
 [75 70 72 80 74]
 [85 90 88 84 91]
 [60 65 62 70 68]
 [95 94 96 89 98]]


**🧮 Interpretation:** Matrix **A** stacks students as rows and subjects as columns — the fundamental structure used for every matrix operation below.

#### (a) Matrix addition and multiplication.


In [34]:
# Matrix Addition
add = A[:5] + A[:5]

# Matrix Multiplication
mul = np.dot(A[:5], A[:5].T)

print("Matrix Addition (A + A):")
print(add)

print("\nMatrix Multiplication (A × Aᵀ):")
print(mul)


Matrix Addition (A + A):
[[184 176 180 156 190]
 [150 140 144 160 148]
 [170 180 176 168 182]
 [120 130 124 140 136]
 [190 188 192 178 196]]

Matrix Multiplication (A × Aᵀ):
[[39417 32810 38857 28740 41904]
 [32810 27585 32465 24146 34989]
 [38857 32465 38406 28474 41377]
 [28740 24146 28474 21193 30656]
 [41904 34989 41377 30656 44602]]


**➕✖️ Insight:**
- **A + A** simply doubles every score.
- **A × Aᵀ** produces the **Gram matrix** — diagonal entries (e.g. 39417 for S01) represent each student’s squared vector length, off-diagonals measure similarity between students. Larger value = more similar score pattern. 🔗

#### (b) Transpose and Inverse (if possible).


In [40]:
# Transpose
transpose = A[:5].T

print("\nTranspose of Matrix:")
print(transpose)

# First 5 students × 5 subjects (5×5 matrix)
B = A[:5, :]

print("\n5×5 Square Matrix:")
print(B)

# Inverse
inv = np.linalg.inv(B[:5])

print("\nInverse Matrix:")
print(inv)


Transpose of Matrix:
[[92 75 85 60 95]
 [88 70 90 65 94]
 [90 72 88 62 96]
 [78 80 84 70 89]
 [95 74 91 68 98]]

5×5 Square Matrix:
[[92 88 90 78 95]
 [75 70 72 80 74]
 [85 90 88 84 91]
 [60 65 62 70 68]
 [95 94 96 89 98]]

Inverse Matrix:
[[ 0.16557626  0.20526113  0.24856813 -0.19328549 -0.4121975 ]
 [ 0.09293007  0.13639554  0.50784416 -0.20200122 -0.52448328]
 [-0.30035316 -0.22037217 -0.41603468  0.10902588  0.76822946]
 [-0.04663482  0.02777715 -0.04047721  0.03803228  0.03542888]
 [ 0.08693093 -0.1391574  -0.28377063  0.23978448  0.12813257]]


**🔁 Interpretation:** The **transpose** flips rows↔columns (subjects become rows). The **inverse** exists because the 5×5 matrix is non-singular — meaning the 5 student vectors are **linearly independent**, spanning the full subject space. ✅

#### (c) Determinant.

In [41]:
# Determinant
det = np.linalg.det(B)

print("\nDeterminant:")
print(det)


Determinant:
-88345.99999999978


**🧾 Insight:** Determinant ≈ **-88,346** (non-zero) ➡️ the matrix is **invertible** and the 5 student score vectors are **linearly independent**. The negative sign indicates the transformation reverses orientation 🔄.

### Part C: Linear Transformations & Geometry



#### Q5. Explain line, plane, and hyperplane with respect to your dataset dimensions.

In [43]:
# LINE (1D)
line = df[["Math"]]

print("Line (1D)")
print(line.head())

# PLANE (2D)
plane = df[["Math", "Physics"]]

print("\nPlane (2D)")
print(plane.head())

# HYPERPLANE (5D)
hyperplane = df[["Math", "Physics", "Chemistry", "English", "Computer"]]

print("\nHyperplane (5D)")
print(hyperplane.head())

Line (1D)
   Math
0    92
1    75
2    85
3    60
4    95

Plane (2D)
   Math  Physics
0    92       88
1    75       70
2    85       90
3    60       65
4    95       94

Hyperplane (5D)
   Math  Physics  Chemistry  English  Computer
0    92       88         90       78        95
1    75       70         72       80        74
2    85       90         88       84        91
3    60       65         62       70        68
4    95       94         96       89        98


**📐 Interpretation:**
- **Line (1D)** → single subject (Math) — a number line.
- **Plane (2D)** → Math vs Physics — a flat 2-axis space.
- **Hyperplane (5D)** → all subjects — impossible to visualise but mathematically valid, where each student is a single point. 🌐


#### Q6. Show how dimensionality increases from 2D → 3D → higher dimensions with hyperplanes.

In [44]:
# 2D
two_d = df[["Math", "Physics"]]

print("2D Shape:", two_d.shape)

# 3D
three_d = df[["Math", "Physics", "Chemistry"]]

print("3D Shape:", three_d.shape)

# 5D
five_d = df[["Math", "Physics", "Chemistry", "English", "Computer"]]

print("5D Shape:", five_d.shape)

2D Shape: (50, 2)
3D Shape: (50, 3)
5D Shape: (50, 5)


**📈 Insight:** As we add subjects, dimensionality grows **2D → 3D → 5D**. Rows stay 50 (students) while columns (features) increase — this is exactly why techniques like **PCA/LDA** later become necessary to compress high-D data. 🔽

### Part D: Eigenvalues & Decomposition


#### Q7.Compute the eigenvalues and eigenvectors of the covariance matrix. 

In [48]:
# Select subject columns
data = df[["Math","Physics","Chemistry","English","Computer"]]

# Covariance Matrix
cov_matrix = np.cov(data.T)

print("Covariance Matrix:")
print(cov_matrix)

# Eigenvalues and Eigenvectors
eigenvalues, eigenvectors = np.linalg.eig(cov_matrix)

print("\nEigenvalues:")
print(eigenvalues)

print("\nEigenvectors:")
print(eigenvectors)

Covariance Matrix:
[[165.15755102 146.76979592 161.13387755  97.91673469 157.60326531]
 [146.76979592 134.34122449 144.41959184  87.52897959 141.88897959]
 [161.13387755 144.41959184 159.8955102   96.41142857 154.91591837]
 [ 97.91673469  87.52897959  96.41142857  61.90040816  93.18693878]
 [157.60326531 141.88897959 154.91591837  93.18693878 153.25265306]]

Eigenvalues:
[665.81258642   4.01127629   2.23109542   1.11025541   1.3821334 ]

Eigenvectors:
[[ 0.49633339 -0.04544677 -0.42279908 -0.47797478  0.58682944]
 [ 0.44613782  0.2875469   0.78391426 -0.31828336 -0.0495173 ]
 [ 0.48825217 -0.08878131 -0.34997911 -0.13602228 -0.78277727]
 [ 0.29678564 -0.86213259  0.27112079  0.2875016   0.11172304]
 [ 0.47764245  0.40508845 -0.10357368  0.75437164  0.16720332]]


**🌟 Key Insight — Eigen Analysis:**
- Eigenvalues: **[665.8, 4.01, 2.23, 1.11, 1.38]**
- The **first eigenvalue dominates** (~98.6% of total variance) ➡️ **one principal direction** explains almost the entire variation among students.
- Interpretation: students mostly differ along a single axis of *overall academic ability* 🎓, not across independent subject strengths.


#### Q8. Perform LU Decomposition of the dataset matrix.  

In [50]:
from scipy.linalg import lu

# Use first 5 students to form a 5×5 square matrix
A = df[["Math","Physics","Chemistry","English","Computer"]].values[:5,:]

# LU Decomposition
P, L, U = lu(A)

print("Matrix A:")
print(A)

print("\nPermutation Matrix (P):")
print(P)

print("\nLower Triangular Matrix (L):")
print(L)

print("\nUpper Triangular Matrix (U):")
print(U)

Matrix A:
[[92 88 90 78 95]
 [75 70 72 80 74]
 [85 90 88 84 91]
 [60 65 62 70 68]
 [95 94 96 89 98]]

Permutation Matrix (P):
[[0. 0. 0. 1. 0.]
 [0. 0. 1. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0.]]

Lower Triangular Matrix (L):
[[ 1.          0.          0.          0.          0.        ]
 [ 0.89473684  1.          0.          0.          0.        ]
 [ 0.78947368 -0.71428571  1.          0.          0.        ]
 [ 0.96842105 -0.51428571  0.825       1.          0.        ]
 [ 0.63157895  0.95535714  0.28125    -0.36253776  1.        ]]

Upper Triangular Matrix (U):
[[ 95.          94.          96.          89.          98.        ]
 [  0.           5.89473684   2.10526316   4.36842105   3.31578947]
 [  0.           0.          -2.28571429  12.85714286  -1.        ]
 [  0.           0.           0.         -16.55         2.625     ]
 [  0.           0.           0.           0.           4.17041163]]


**🔺 Interpretation (LU):** Matrix **A = P · L · U** — decomposed into Permutation, Lower- and Upper-triangular parts. This factorisation makes solving linear systems and computing determinants much faster ⚡ and is numerically stable due to pivoting (P).


#### Q9. Perform Singular Value Decomposition (SVD) and explain its role in dimensionality reduction. 

In [62]:
# Dataset matrix
A = df[["Math","Physics","Chemistry","English","Computer"]].values

# Singular Value Decomposition
U, S, VT = np.linalg.svd(A)

print("Matrix U:")
print(U)

print("\nSingular Values:")
print(S)

print("\nMatrix Vᵀ:")
print(VT)

Matrix U:
[[-1.57451701e-01  3.43613035e-01 -1.13853004e-03 ... -3.57528141e-02
  -1.74025795e-01 -4.09819006e-02]
 [-1.31748687e-01 -2.07529076e-01 -4.21012845e-01 ... -4.32602536e-02
   5.00017973e-02 -3.49367919e-02]
 [-1.55635883e-01  6.38275456e-02  2.51737718e-01 ...  1.66082684e-02
   9.40675642e-02 -1.87020631e-02]
 ...
 [-1.05146676e-01 -2.00355244e-01  1.03633573e-03 ...  9.24663744e-01
  -1.34340144e-02 -2.45247888e-02]
 [-1.47802799e-01  3.28707077e-02  8.78143384e-02 ... -1.86200879e-05
   9.59041472e-01 -2.50510519e-04]
 [-1.19700647e-01 -1.61211246e-01 -2.81603476e-02 ... -3.80554575e-02
  -1.81192417e-02  9.59060971e-01]]

Singular Values:
[1258.92167942   31.77134214   11.64442792    8.37137639    8.00713808]

Matrix Vᵀ:
[[-0.44288462 -0.44742622 -0.44539017 -0.44107184 -0.45907104]
 [ 0.3421857   0.04641843  0.27793354 -0.87480771  0.19549604]
 [-0.42869615  0.72581502 -0.40927765 -0.19439161  0.29002738]
 [-0.3590266   0.36368548  0.63485381 -0.04848279 -0.57744508]


**🧩 Insight (SVD):** SVD factorises the data as **A = U · Σ · Vᵀ**.
- **Singular values (S)** rank the importance of each direction.
- Keeping only the top-k singular values gives an optimal **low-rank approximation** — the mathematical foundation of **dimensionality reduction, compression, and noise removal** 🎯.

### Part E: Dimensionality Reduction



#### Q10. Apply Principal Component Analysis (PCA) to reduce the dataset from multiple subjects to 2 dimensions.

In [60]:
from sklearn.decomposition import PCA
pca = PCA(n_components=2)
reduced = pca.fit_transform(A)

print("Reduced Data (2D):\n", reduced[:5])

Reduced Data (2D):
 [[ 23.51923004  -6.90450367]
 [-11.17437768   6.13180447]
 [ 19.8308116   -1.18213818]
 [-31.56630048  -0.19077088]
 [ 35.31213953   0.3074362 ]]


**🎨 Insight (PCA → 2D):** The 5-D subject space is compressed to 2 components while preserving maximum variance. Notice how *Above Average* students get **large positive PC1** values (e.g. 23.5, 19.8, 35.3) and *Below Average* students get **negative PC1** values (-11.2, -31.6). PC1 clearly encodes **overall performance**. 📉📈


#### Q11. Apply Linear Discriminant Analysis (LDA) to classify students into “Above Average” and “Below Average” categories.

In [59]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

# Features
X = df[["Math", "Physics", "Chemistry", "English", "Computer"]].values

# Convert labels to numbers
y = df["Category"].map({
    "Above Average": 1,
    "Below Average": 0
})

# Create LDA object
lda = LinearDiscriminantAnalysis(n_components=1)

# Transform the data
lda_result = lda.fit_transform(X, y)

print("LDA Result:")
print(lda_result[:5])

LDA Result:
[[ 4.27945532]
 [-2.4641946 ]
 [ 2.39589138]
 [-1.79223201]
 [ 3.5004873 ]]


**🧭 Insight (LDA):** LDA projects the 5-D data onto **one axis that best separates the two classes**. Positive LDA scores → *Above Average* 🌟, negative → *Below Average* 📉. The clean sign-split (e.g. +4.28, +2.40, +3.50 vs −2.46, −1.79) shows the two groups are **linearly separable** — LDA is an excellent classifier for this dataset. ✅

## 🏁 Overall Conclusion & Key Takeaways

### 📌 Dataset Summary
- **50 students × 5 subjects** — clean data, no missing values ✅
- Average performance is highest in **Computer 💻** and most consistent in **English 📖**

### 🔗 Vector & Matrix Findings
- Student score vectors are **highly aligned** (angle ≈ 5.7° between S01 & S02) — students differ mostly in **magnitude of performance**, not in *subject profile*.
- The 5×5 score matrix is **invertible** (det ≠ 0) → student vectors are **linearly independent**.
- The **Gram matrix (A·Aᵀ)** gives a natural student-similarity measure.

### 🌟 Eigen / SVD / PCA Insights
- The **first eigenvalue captures ≈ 98.6% of total variance** — a single latent factor (**overall academic ability 🎓**) drives the dataset.
- Both **PCA** and **LDA** confirm this: **PC1** and the **LDA axis** cleanly separate *Above Average* from *Below Average* students.

### 🧭 Practical Implications
- The dataset is effectively **1-dimensional in nature** — one score summarises a student well.
- **LDA is a reliable classifier** for the *Above / Below Average* label because the two classes are **linearly separable**.
- Techniques like **PCA & SVD** are ideal for **compressing** or **visualising** this multi-subject data without significant information loss. 🎯

### ✅ Final Verdict
The mathematical foundation — **vectors, matrices, eigen decomposition, SVD, PCA and LDA** — collectively reveal that student performance across subjects is **strongly correlated and low-dimensional in structure**, making the data both **well-behaved and highly predictable**. 🚀
